### PR TITLE
Add support for JWKS signature verification and /keys endpoint

### PR DESCRIPTION
**Pull request by sdelamo**
Close https://github.com/micronaut-projects/micronaut-core/issues/852

Close https://github.com/micronaut-projects/micronaut-core/issues/851

### CODE REVIEW

File: security-jwt/src/main/java/io/micronaut/security/token/jwt/endpoints/KeysController.java [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+    public KeysController(Collection<JwkProvider> jwkProviders) {
+        this.jwkProviders = jwkProviders;
+    }
+
+    /**
+     *
+     * @return a JSON Web Key Set (JWKS) payload.
+     */
+    @Get
+    public JSONObject keys() {
```
Comment (jameskleeh) [inline security-jwt/src/main/java/io/micronaut/security/token/jwt/endpoints/KeysController.java]: This should be reactive

File: security-jwt/src/main/java/io/micronaut/security/token/jwt/endpoints/KeysControllerConfiguration.java [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+
+import io.micronaut.core.util.Toggleable;
+
+/**
+ * Encapsulates the configuration of {@link KeysController}.
+ *
+ * @author Sergio del Amo
+ * @since 1.1.0
+ */
+public interface KeysControllerConfiguration extends Toggleable  {
```
Comment (jameskleeh) [inline security-jwt/src/main/java/io/micronaut/security/token/jwt/endpoints/KeysControllerConfiguration.java]: Should probably add the setters here

File: security-jwt/src/main/java/io/micronaut/security/token/jwt/signature/jwks/JwksSignatureConfiguration.java [hunk 3]
```diff
@@ -0,0 +1,267 @@
...
+    private KeyType keyType;
+    private String url;
+
+    /**
+     *
+     * @param jwksSignatureConfigurationProperties JSON Web Key Set configuration.
+     */
+    public JwksSignatureConfiguration(JwksSignatureConfigurationProperties jwksSignatureConfigurationProperties) {
+        if (LOG.isDebugEnabled()) {
+            LOG.error("JWT validation URL: {}", jwksSignatureConfigurationProperties.getUrl());
```
Comment (jameskleeh) [inline security-jwt/src/main/java/io/micronaut/security/token/jwt/signature/jwks/JwksSignatureConfiguration.java]: logging error inside `if (debugEnabled)`

File: security-jwt/src/main/java/io/micronaut/security/token/jwt/signature/jwks/JwksSignatureConfiguration.java [hunk 4]
```diff
@@ -0,0 +1,267 @@
...
+        }
+        return new RSASSAVerifier(rsaPublicKey);
+    }
+
+    /**
+     *
+     * @param ecKey A ECKey
+     * @return a JWSVerifier for a ECKey.
+     */
+    protected JWSVerifier jwsVerifierForECKey(ECKey ecKey) {
```
Comment (jameskleeh) [inline security-jwt/src/main/java/io/micronaut/security/token/jwt/signature/jwks/JwksSignatureConfiguration.java]: Couldn't the 2 try catches be combined? The exception that gets logged should inform the user which step of the process failed

File: security-jwt/src/main/java/io/micronaut/security/token/jwt/signature/jwks/JwksSignatureFactory.java [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+import io.micronaut.context.annotation.Factory;
+
+/**
+ * Instantiates a {@link io.micronaut.security.token.jwt.signature.jwks.JwksSignatureConfiguration} per {@link JwksSignatureConfigurationProperties} bean.
+ *
+ * @author Sergio del Amo
+ * @since 1.1.0
+ */
+@Factory
+public class JwksSignatureFactory {
```
Comment (jameskleeh) [inline security-jwt/src/main/java/io/micronaut/security/token/jwt/signature/jwks/JwksSignatureFactory.java]: You don't need a factory for this. You can add `@EachBean` to the `JwksSignatureConfiguration` class

File: security-jwt/src/main/java/io/micronaut/security/token/jwt/signature/rsa/RSASignatureGenerator.java [hunk 6]
```diff
@@ -64,7 +69,8 @@ public SignedJWT sign(JWTClaimsSet claims) throws JOSEException {
      */
     protected SignedJWT signWithPrivateKey(JWTClaimsSet claims, @NotNull RSAPrivateKey privateKey) throws JOSEException {
         final JWSSigner signer = new RSASSASigner(privateKey);
-        final SignedJWT signedJWT = new SignedJWT(new JWSHeader(algorithm), claims);
+        JWSHeader jwsHeader = keyId == null ? new JWSHeader(algorithm) : new JWSHeader.Builder(algorithm).keyID(keyId).build();
```
Comment (jameskleeh) [inline security-jwt/src/main/java/io/micronaut/security/token/jwt/signature/rsa/RSASignatureGenerator.java]: Could you just use the builder here? Setting the `keyID` to null should have no impact

File: security-jwt/src/main/java/io/micronaut/security/token/jwt/endpoints/KeysControllerConfiguration.java (hunk 2 above)
Comment (sdelamo) [reply security-jwt/src/main/java/io/micronaut/security/token/jwt/endpoints/KeysControllerConfiguration.java]: I think it is fine without setter

**Comment by jameskleeh** [comment]
This is missing tests for JwksSignature. Other than that good to go

**Comment by sdelamo** [comment]
@jameskleeh `JwksSpec` was testing the golden path for `JwsSiganture`.

 I have added additional tests to verify the code behaves as expected when a JWKS endpoint throws a 500 or if you attempt to verify a signed JWT which uses the same algorithm as the one used by the `JwsSiganture` but which was signed with a different key.

**Pull request merged**